services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/Authorizer.kt

```diff
@@ -0,0 +1,37 @@
+package fi.hel.haitaton.hanke.permissions
+
+import fi.hel.haitaton.hanke.HankeNotFoundException
+import fi.hel.haitaton.hanke.HankeRepository
+import fi.hel.haitaton.hanke.currentUserId
+import org.springframework.stereotype.Component
+import org.springframework.transaction.annotation.Transactional
+
+abstract class Authorizer(
+    private val permissionService: PermissionService,
+    private val hankeRepository: HankeRepository? = null
+) {
+    fun authorize(hankeId: Int?, permissionCode: PermissionCode, ex: () -> Exception) {
+        hankeId
+            ?.let { permissionService.hasPermission(it, currentUserId(), permissionCode) }
+            .let { if (it != true) throw ex() }
+    }
+
+    internal fun authorizeHankeTunnus(
+        hankeTunnus: String,
+        permissionCode: PermissionCode
+    ): Boolean {
+        val hankeIdentifier = hankeRepository!!.findOneByHankeTunnus(hankeTunnus)
+        authorize(hankeIdentifier?.id, permissionCode) { HankeNotFoundException(hankeTunnus) }
+        return true
+    }
+
+    @Transactional(readOnly = false)
+    open fun authorizeHankeTunnus(hankeTunnus: String, permissionCode: String): Boolean =
+        authorizeHankeTunnus(hankeTunnus, PermissionCode.valueOf(permissionCode))
+}
+
+@Component
+class HankeAuthorizer(
```

Comment:
Is there a reason why only HankeAuthorizer is placed here? Would it be better to either have this in a separate file or move application and hankekayttajaauthorizers here too?

Looking at this it seemed (to me at least) that there is only one implementation.

Reply:
Moved to hanke-package as a separate file.

Some kind of package-reconfiguration might be in order for Hanke-related classes. They are weirdly in the root package of the project, while other concepts are in their own packages.

Reply:
That seems to be because of fi.hel.haitaton.hanke being the root (would lead to hanke.hanke structure). Maybe Hanke stuff could be moved to domain package or the root should be just fi.hel.haitaton and then we could have a separate hanke package (I would prefer having haitaton as root). Maybe not as part of this pr.